### ISSUE TITLE
Add new BDO format reader for SHIELD-HIT12A

### ISSUE THREAD

**Issue by grzanka** [DataMedSci/pymchelper#336]

**Comment by nbassler** [comment]
this will be needed in not too far future..

**Comment by nbassler** [comment]
@fkalholm 
Suggestion for test `.bdo` test files:

125 MeV proton beam on 20 x 20 x 20 cm³ water target along z.

Test most simple cases:
- `00_iddc_1d.bdo`  : dose scoring 20 x 20 x 20 cm³, 5 slices along z (4 cm each bin)
- `01_iddc_2d.bdo`  : dose scoring 20 x 20 x 20 cm³, 5 slices along z and y (4 cm each bin)
- `02_iddc_3d.bdo`  : dose scoring 20 x 20 x 20 cm³, 5 slices along z, y and x (4 cm each bin)
- `03_iddc_1d.bdo`  : fluence scoring 20 x 20 x 20 cm³, 5 slices along z (4 cm each bin), filtered on primary particles.

Test differential bins:
- `10_diff_prm.bdo`  : scoring 20 x 20 x 1 cm³, 1 slice at surface. Differential spectrum (dPhi/dE), linear steps 0 - 200 MeV in 20 steps. Filtered on primary protons only. 
- `11_diff_z1a.bdo`  : scoring 20 x 20 x 1 cm³, 1 slice at surface. Differential spectrum (dPhi/dE), linear steps 0 - 200 MeV in 20 steps. 3 pages. Page 1: all protons (z=1 a=1, inc. secondary protons), Page 2: all deuterons, Page 3: all tritons.

Test log binning:
- `20_diff_prm.bdo` : same as above, but in log steps, lower limit 0.1 MeV
- `21_diff_z1a.bdo` : same as above, but in log steps, lower limit 0.1 MeV

Test differential energy type:
- `30_diff_prm.bdo` : same as `10_diff_prm.bdo`, but energy in E/amu.
 

Interesting observations @grzanka :
-  the current .spc format is not able do discriminate between primary and secondary protons. This means we need to extend the .spc format with new tags?

**Comment by fkalholm** [comment]
I added the new bdo-files, along with corresponding txt-files for convenience now. 
@nbassler I could not get the EAMU unit to work (should be a very small difference compared to ENUC, but all fluence ended up in the first bin). Maybe there is something I misunderstand, or it could be a bug.

**Comment by nbassler** [comment]
```
/*
   Concept:

   Each SH12A run has unique settings, such as SH12A version, beam config. These go into the BDO_HEADER.
   Each SH12A run has one or more estimators.
   Each estimator holds exactly one geometry.
   Each estimator has a unique filename, and will therefore go into a new .bdo file.
   Each estimator/bdo file contains one or more pages.
   Each page holds exactly one detector.

   Output:
   Everything is padded to modulo 8 bytes.

   First 8 bytes: a 6 byte ASCII magic number + a 2-byte ASCII string indicating big or little endian.

   Then follows a series of tokens.
   A token consists of a tag and a payload.
   The tag is always fixed in size.
   The payload size varies, and the size is specified in the tag.

   A tag holds
   1) tag_id (8-bytes)
   2) payload descriptor : 8-byte ASCII descriptor for a single element of the payload. E.g. "<8f" for 8-byte floats.
   3) payload length : size_t length, the number of elements (not bytes) in the payload.

   Token = [tag-id (8-bytes)] [payload descriptor (8-bytes)] [payload length (8-bytes)] [payload (variable size)]

   The payload has variable length but is always 8-byte padded.
   ASCII strings in the payload will always be null-byte terminated with at least one null byte.
   Integers are always 8-byte integers (long long).
   Floats are always 8-byte floats (double).
   Therefore, in case of integers and floats, the number of elements in a payload is always the bytelength / 8.
 */
```


Header files for documentation of new file format (this may not be the final version)

sh_file_format.h
```
#ifndef _SH_FILE_FORMAT
#define _SH_FILE_FORMAT

/* Shared defines for multiple bdo subformats.
   This file should be merged with sh_bdo.h, once the obsolete sh_detect.h and sh_detect.c is removed.
 */

/* defines for output file formats */
#define SHBDO_MAGIC_NUMBER "xSH12A"    /* 6-letter ASCII code to identify detector files */

#define SH_FILE_FORMAT_ID_UNDEFINED  -1
#define SH_FILE_FORMAT_ID_BIN2010     0    /* Old binary format from 2010, the first version David wrote back then */
#define SH_FILE_FORMAT_ID_BDO2016     1    /* year 2016 .bdo file format, now with proper tags, used from SH 0.6.0 */
#define SH_FILE_FORMAT_ID_BDO2019     2    /* year 2019 .bdo style, introduced June 2019 */
#define SH_FILE_FORMAT_ID_ASCII       3    /* raw text format */
#define SH_FILE_FORMAT_ID_CSV         4    /* comma separated file format */

#endif  /* !_SH_FILE_FORMAT */
```

sh_bdo.h
```
#ifndef _SH_BDO
#define _SH_BDO

#include <stdlib.h>
#include <stdint.h>

/* datatype strings as in https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html */
#define SHBDO_PL_TYPE_NONE    "V"   /* raw data (void) */
#define SHBDO_PL_TYPE_CHAR    "S"   /* unsigned char  */
#define SHBDO_PL_TYPE_LLUINT  "u8"  /* long long unsigned int */
#define SHBDO_PL_TYPE_LLSINT  "i8"  /* long long signed int */
#define SHBDO_PL_TYPE_DOUBLE  "f8"  /* 8-byte double float */

enum bdotags {
    /*
       Hex values are used for better regognition in binary files, should they be inspected by humans.
     */
    /* Group 0x0000 - 0x00FF : Miscellaneous info */
    SHBDO_SHVERSION = 0x0000,   /* [char*] version of SHIELD-HIT12A */
    SHBDO_SHBUILDDATE,          /* [char*] date of build */
    SHBDO_FILEDATE,             /* [char*] bdo file creation date, RFC 2822 compliant */
    SHBDO_USER,                 /* [char*] optional login name */
    SHBDO_HOST,                 /* [char*] optional host where this file was created */
    SHBDO_FORMAT,               /* [int]   optional ID describing which flavour the format is, to help .bdo parsers */

    /* Group 0xCC00 - 0xCCFF : Configuration */
    SHBDO_DELE = 0xCC00,
    SHBDO_DEMIN,
    SHBDO_ITYPST,
    SHBDO_ITYPMS,
    SHBDO_OLN,
    SHBDO_INUCRE,
    SHBDO_IEMTRANS,
    SHBDO_IEXTSPEC,
    SHBDO_INTRFAST,
    SHBDO_INTRSLOW,
    SHBDO_APZLSCL,
    SHBDO_IOFFSET,
    SHBDO_IRIFIMC,
    SHBDO_IRIFITRANS,
    SHBDO_IRIFIZONE,
    SHBDO_EXT_NPROJ,
    SHBDO_EXT_PTVDOSE,
    SHBDO_IXFIRS,

    /* # Group 0xCE00 - 0xCEFF : CT specific tags */
    SHBDO_CT_ANG = 0xCE00,   /* holds two doubles with the couch and gantry angle */
    SHBDO_CT_ICNT,           /* holds three */
    SHBDO_CT_LEN,            /* holds three */

    /* Group 0xEE00 - 0xEEFF : Estimator specific tags */
    SHBDO_EST_GEOTYP = 0xEE00,
    SHBDO_EST_PAGES,             /* number of detectors / pages for this estimator */

    /*  ------------- NEW STUFF HERE -------------------- */
    SHBDO_EST_PAG_DIF_TYPE,
    SHBDO_EST_PAG_DIF_START,
    SHBDO_EST_PAG_DIF_STOP,
    SHBDO_EST_PAG_DIF_SIZE,

    /* Group 0xEF00 - 0xEFFF : Filter data attached to estimator */
    SHBDO_EST_PAG_FILTER_NAME = 0xEF00,
    SHBDO_EST_PAG_FILTER_EMIN,   /* lower energy threshold, emin */

    /* Group 0xE000 - 0xE0FF : Geometry data attached to estimator */
    SHBDO_EST_GEO_TYPE = 0xE000,
    SHBDO_EST_GEO_P,       /* start values, e.g xmin, ymin, zmin */
    SHBDO_EST_GEO_Q,       /* stop values, e.g xmax, ymax, zmax */
    SHBDO_EST_GEO_N,       /* number of bins */
    SHBDO_EST_GEO_VOL,     /* volume size in cm3 ... WARNING: may be a list in some future */
    SHBDO_EST_GEO_ZONES,   /* single GEMCA zone, or list of zones */
    SHBDO_EST_GEO_NAME,    /* Usergiven Name of this geometry */
    /*  ------------- END OF NEW STUFF -------------------- */

    /* Group 0xDD00 - 0xDDFF : Detector / page specific tags */
    SHBDO_DET_GEOTYP = 0xDD00,  /* may differ from SHBDO_EST_GEOTYP in case of SPC */    /* deprecated */
    SHBDO_DET_NBIN,       /* IDET(1-3) (len=3) number of bins X,Y,Z */                   /* deprecated */
    SHBDO_DET_PART,       /* IDET(4) particle type which was scored */
    SHBDO_DET_DTYPE,        /* IDET(5) detector type */
    SHBDO_DET_PARTZ,      /* IDET(6) */
    SHBDO_DET_PARTA,        /* IDET(7) */
    SHBDO_DET_DMAT,       /* IDET(8) */
    SHBDO_DET_NBINE,          /* IDET(9) */    /* deprecated */
    SHBDO_DET_DIFFTYPE,     /* IDET(10) */   /* deprecated */
    SHBDO_DET_ZONESTART,  /* IDET(11) */   /* deprecated */
    SHBDO_DET_DSIZE,      /* IDET(12) */   /* deprecated */
    SHBDO_DET_DSIZEXYZ,   /* IDET(13) */   /* deprecated */

    SHBDO_DET_XYZ_START,    /* DET(1-3) */   /* deprecated */
    SHBDO_DET_XYZ_STOP,     /* DET(4-6) */   /* deprecated */
    SHBDO_DET_DIF_START,    /* DET(7) */     /* deprecated */
    SHBDO_DET_DIF_STOP,     /* DET(8) */     /* deprecated */
    SHBDO_DET_VOXVOL,         /* DET(9) */     /* deprecated */
    SHBDO_DET_THRESH,     /* DET(10) */    /* deprecated */

    SHBDO_DET_DATA = 0xDDBB,    /* data block */

    /* Group 0xAA00 - 0xAAFF : Runtime variables */
    SHBDO_RT_NSTAT = 0xAA00,    /* number of actually simulated particles */
    SHBDO_RT_TIME,            /* [usignend long int] optional runtime in seconds */

    /* Group 0xFFCC - 0xFFFF : Diagnostics, may be ignored by readers. */
    SHBDO_COMMENT = 0xFFCC,
    SHBDO_DEBUG,                /* 0xFFCD */
    SHBDO_ERROR                /* 0xFFCE */
};

struct SHBDO_TAG {   /* 8 bytes */
    uint64_t tag;
    /* pltype[8]: payload datatype str, numpy style as in
       https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html
       see "Array-protocol type strings"
     */
    char pltype[8];  /* TODO: find another solution - cannot be larger than 8, else struct will be padded */
    size_t len;         /* number of elements in payload (NOT number of bytes) */
};

#endif  /* !_SH_BDO */
```

**Comment by nbassler** [comment]
> @nbassler I could not get the EAMU unit to work for the 30-file (should be a very small difference compared to ENUC, but all fluence ended up in the first bin). Maybe there is something I misunderstand, or it could be a bug.

Is your bin resolution high enough that you can detect the difference?

**Comment by fkalholm** [comment]
If it was the same numbers, that could be it. But the issue I have is that for ENUC, all fluence comes in bin 11 and 12 (corresponding to the ~125 MeV). For EAMU, everything elds up in bin 0

**Comment by nbassler** [comment]
ah, then I misunderstood  you entirely. This is a SH12A issue. Should be reported there.

**Comment by nbassler** [comment]
@grzanka 
Just realized that
- currently the .bdo writer still assumes filters attached on estimator level. We agreed on page level. 
- geometries are also attached on estimator level, which may lead to data-duplication.
- pages are still not properly handled.

Any preference for a decent and lasting solution?

Proposed structure (this is a bit more classy):
```
----------------------------
- header data
- geometry MyMsh
- geometry MyCyl
- geometry ...
- estimator 1
     - geoname MyMSH
     - page 1 (Fluence)
         - data
     - page 2 (Fluence+filter)
         - data
     - page 3 (diff Fluence+filter2)
         - data
- estimator 2
     - geoname MyMSH
     - page 1 (Dose)
         - data
...
----------------------------
```

Old structure:
```
----------------------------
- header data
- estimator 1
     - geometry MyMsh
     - page 1 (Fluence)
         - data
     - page 2 (Fluence+filter)
         - data
     - page 3 (diff Fluence+filter2)
         - data
- estimator 2
     - geometry MyMsh
     - page 1 (Dose)
         - data
...
----------------------------

**Comment by grzanka** [comment]
Fixed in #348
